k2/python/k2/mwer_loss.py

```diff
@@ -0,0 +1,436 @@
+# Copyright (c)  2022  Xiaomi Corporation (authors: Liyong Guo)
+
+import torch
+import torch.nn as nn
+import k2
+from typing import Dict, Iterable, List, Optional, TextIO, Tuple, Union
+
+# Copied from icefall.
+# TODO(liyong) simplify this.
+def get_texts(
+    best_paths: k2.Fsa, return_ragged: bool = False
+) -> Union[List[List[int]], k2.RaggedTensor]:
+    """Extract the texts (as word IDs) from the best-path FSAs.
+    Args:
+      best_paths:
+        A k2.Fsa with best_paths.arcs.num_axes() == 3, i.e.
+        containing multiple FSAs, which is expected to be the result
+        of k2.shortest_path (otherwise the returned values won't
+        be meaningful).
+      return_ragged:
+        True to return a ragged tensor with two axes [utt][word_id].
+        False to return a list-of-list word IDs.
+    Returns:
+      Returns a list of lists of int, containing the label sequences we
+      decoded.
+    """
+    if isinstance(best_paths.aux_labels, k2.RaggedTensor):
+        # remove 0's and -1's.
+        aux_labels = best_paths.aux_labels.remove_values_leq(0)
+        # TODO: change arcs.shape() to arcs.shape
+        aux_shape = best_paths.arcs.shape().compose(aux_labels.shape)
+
+        # remove the states and arcs axes.
+        aux_shape = aux_shape.remove_axis(1)
+        aux_shape = aux_shape.remove_axis(1)
+        aux_labels = k2.RaggedTensor(aux_shape, aux_labels.values)
+    else:
+        # remove axis corresponding to states.
+        aux_shape = best_paths.arcs.shape().remove_axis(1)
+        aux_labels = k2.RaggedTensor(aux_shape, best_paths.aux_labels)
+        # remove 0's and -1's.
+        aux_labels = aux_labels.remove_values_leq(0)
+
+    assert aux_labels.num_axes == 2
+    if return_ragged:
+        return aux_labels
+    else:
+        return aux_labels.tolist()
+
+
+# Copied from icefall.
+# TODO(liyong) simplify this.
+class Nbest(object):
```

Comment:
If this is going to be used it should probably be in a different file.   But are you sure it's not already in k2?

Reply:
Yes, class `Nbest` is also exists in k2. 
But it has fewer functions than that in icefall, i.e. `from_lattice` and `build_levenshtein_graphs`. So I just copied the icefall one here. 
Now in the latest pr, this two needed functions  are added into the k2.Nbest.